onnxruntime/core/providers/cuda/reduction/reduction_ops.h

```diff
@@ -2,32 +2,50 @@
 // Licensed under the MIT License.
 
 #pragma once
+#include "core/common/optional.h"
 #include "core/providers/cuda/cuda_common.h"
 #include "core/providers/cpu/reduction/reduction_ops.h"
+#include "core/providers/cuda/reduction/reduction_functions.h"
 
 namespace onnxruntime {
 namespace cuda {
 
 template <bool allow_multi_axes>
 class ReduceKernel : public CudaKernel, public ReduceKernelBase<allow_multi_axes> {
  protected:
-  ReduceKernel(const OpKernelInfo& info) : CudaKernel(info),
-                                           ReduceKernelBase<allow_multi_axes>(info),
-                                           calculate_log_(false),
-                                           calculate_sqt_(false),
-                                           log_sum_exp_(false) {}
+  ReduceKernel(
+      const OpKernelInfo& info,
+      optional<int64_t> keep_dims_override = {})
+      : CudaKernel(info),
+        ReduceKernelBase<allow_multi_axes>(info, keep_dims_override),
+        calculate_log_(false),
+        calculate_sqt_(false),
+        log_sum_exp_(false),
+        fast_reduction_(false) {}
 
   // Only Max Min need to set ReduceTensorIndices CUDNN_REDUCE_TENSOR_FLATTENED_INDICES as per cudnn library manual
   // Only Max Min will have indices output, need to set the indices to nullptr for other ops
   template <typename T, cudnnReduceTensorIndices_t ReduceTensorIndices = CUDNN_REDUCE_TENSOR_NO_INDICES>
   Status ComputeImpl(OpKernelContext* ctx, cudnnReduceTensorOp_t cudnnReduceOp) const;
 
+  template <typename T, typename OutT, cudnnReduceTensorIndices_t ReduceTensorIndices>
+  Status ReduceKernelShared(
```

Comment:
I don't see this getting used anywhere or even defined. Do we need it?

Reply:
per comments in .cc  - ReduceKernelShared() is still used by some other training classes though it's not used here